spec/jobs/movements_on_date_job_spec.rb

```diff
@@ -1,38 +1,76 @@
+  let(:nomis_offender_id) { nomis_offender.fetch(:offenderNo) }
+  let(:nomis_offender) { build(:nomis_offender, agencyId: to_prison_code) }
+  let(:case_info) { create(:case_information) }
+  let!(:offender_record) { create(:offender, nomis_offender_id: nomis_offender_id, case_information: case_info) }
+  let(:today) { Time.zone.today }
+  let(:yesterday) { today - 1.day }
+  let(:pom) { build(:pom) }
+  let!(:allocation) { create(:allocation_history, nomis_offender_id: nomis_offender_id, primary_pom_nomis_id: pom.staff_id, prison: from_prison_code) }
+  let(:mail_params) {
+    {
+      email: pom.email_address,
+      pom_name: pom.first_name.titleize,
+      offender_name: "#{nomis_offender.fetch(:lastName)}, #{nomis_offender.fetch(:firstName)}".titleize,
+      nomis_offender_id: nomis_offender_id,
+      prison_name: Prison.find(from_prison_code).name,
+      url: "http://localhost:3000/prisons/#{from_prison_code}/staff/#{pom.staff_id}/caseload"
+    }
+  }
 
   before do
-    stub_auth_token
-  end
+    stub_offender(nomis_offender)
+    stub_poms(from_prison_code, [pom])
+
+    # Stub offender movements from yesterday
+    stub_request(:get, "#{ApiHelper::T3}/movements?fromDateTime=#{yesterday - 1.year}T00:00&movementDate=#{yesterday}").
+      to_return(body: [movement].to_json)
 
-  it 'deallocates' do
-    allow(OffenderService).to receive(:get_offender).and_return(build(:hmpps_api_offender))
+    # Expect POM to be emailed about the deallocation
+    expect(PomMailer).to receive(:offender_deallocated).with(mail_params).and_return(double(deliver_later: nil))
+  end
 
-    stub_request(:get, "#{ApiHelper::T3}/movements?fromDateTime=2018-06-30T00:00&movementDate=2019-06-30").
-      to_return(body: [attributes_for(:movement,
-                                      offenderNo: nomis_offender_id,
-         fromAgency: "MDI",
-         toAgency: "WEI",
-         movementType: "ADM",
-         directionCode: "IN")].to_json)
+  context 'when the offender has been released from prison' do
+    let(:from_prison_code) { create(:prison).code }
+    let(:to_prison_code) { 'OUT' } # "OUT" is the prison code NOMIS uses to mean "not in prison"
 
-    pom = build(:pom, staffId: 485926)
-    stub_poms('MDI', [pom])
-    stub_pom pom
-    stub_offenders_for_prison('MDI', [])
+    let(:movement) {
+      attributes_for(:movement,
+                     offenderNo: nomis_offender_id,
+                     movementType: 'REL',
+                     directionCode: 'OUT',
+                     fromAgency: from_prison_code,
+                     toAgency: to_prison_code)
+    }
 
-    expect(alloc.primary_pom_nomis_id).not_to be_nil
-    expect(alloc.secondary_pom_nomis_id).not_to be_nil
+    it 'deallocates and emails the POM' do
+      expect(allocation).to be_active
+      described_class.perform_now(today.to_s)
+      allocation.reload
+      expect(allocation).not_to be_active
+    end
+  end
```

Comment:
It does feel like allocationHistory is being tested inside MovementOnDateJob. It feels a bit like an integration tests because you are testing how the actions of one effect the other. Is this just the nature of testing Jobs or is this because they are quite tightly coupled?

Reply:
@Katy600 yes – you're spot on. I initially wanted to write this as a feature spec – but then I realised that it doesn't really make sense to test this as a browser based user journey, since it's not! Then I spotted that there was already a job spec which half-covered the existing functionality. So my approach was to build out that spec instead of create a new one.

Either way: it's definitely a high level test. From a business point of view, we want to be sure that POMs are deallocated (and notified by email) when offenders are released or transferred out of that prison – so I think it's useful to have a test which broadly covers that.

The implementation itself is spread across different classes (as you've already called out: AllocationHistory does the 'meat' of the work). However this job is actually the only thing that uses that functionality – so in terms of implementation code, it could just as easily be shuffled over into the job itself. (And it might make more sense to do so, because it'd remove some noise from the AllocationHistory model.)

I'm not sure if that really answers anything – but I'm just agreeing with what you've said, and explaining my thinking a bit!

Reply:
@ollietreend Thanks Ollie. Yes that makes so much sense. It does feel it would make more sense to be in the job now that you have mentioned it. As if it is not being used by any thing else then it feels like it shouldn't be the responsibility of the AllocationHistory model to home it.